### PR TITLE
Fix #63, support data attributes.

### LINKED ISSUE #63: data attributes
there is no way to set them. 
```purescript
attr (attributeName "data-foo") "bar" 
```
doesn't work

### PR DESCRIPTION
**Pull request by paf31**
@jdegoes @cryogenian @garyb Could you please review?

Details here:

https://github.com/Matt-Esch/virtual-dom/blob/7cd99a160f8d7c9953e71e0b26a740dae40e55fc/docs/vnode.md#custom-attributes-data-

### CODE REVIEW

**Comment by garyb** [comment]
Is it possible for `attributes` to be set in `props`? If so, that may overwrite `data-` properties set this way - we'd need need to merge a `props.attributes` and `props1.attributes` instead.

File: src/Halogen/Internal/VirtualDOM.purs [hunk 1]
```diff
@@ -128,7 +128,19 @@ foreign import vnode
   \  return function(props) {\
   \    return function(children) {\
   \      var VirtualNode = require('virtual-dom/vnode/vnode');\
-  \      return new VirtualNode(name, props, children);\
+  \      var props1 = {\
+  \        attributes: {}\
+  \      };\
+  \      for (var key in props) {\
```
Comment (garyb) [inline src/Halogen/Internal/VirtualDOM.purs]: We might want to iterate over `Object.keys(props)` here instead, as then you don't need the `hasOwnProperty` call for every key.

**Comment by paf31** [comment]
The only way that's possible right now is if the user creates a custom attribute, of type `{ attributes: _ }`, which I consider to be a misuse anyway. 

I actually tried a different approach in which the `props` constructor branched depending on the key, but then the `combineProps` function got a bit confusing. 

I guess it depends what we want the `Props` type to represent. If it's just a map for internal use, I think it's probably fine, as long as people stick to the API. If we want it to represent valid `virtual-dom` property objects, ready to be passed to `vnode`, then this obviously won't work.

File: src/Halogen/Internal/VirtualDOM.purs (hunk 1 above)
Comment (paf31) [inline src/Halogen/Internal/VirtualDOM.purs]: That would need a polyfill for ES3 right? Which is fine obviously, but would need documenting.

Comment (garyb) [inline src/Halogen/Internal/VirtualDOM.purs]: Yep. I thought we were kind of heading that way for all PS code anyway. I know there's ES5 stuff that crept into `Array` already at least.

Comment (paf31) [inline src/Halogen/Internal/VirtualDOM.purs]: Ok, sounds good.

Comment (paf31) [inline src/Halogen/Internal/VirtualDOM.purs]: So would that be `Object.keys(props).forEach(...)` then?

Comment (garyb) [inline src/Halogen/Internal/VirtualDOM.purs]: For optimum speed it'd still be a for loop, just looping over the array of extracted keys instead. `forEach` is probably plenty fast enough though.

**Comment by jdegoes** [comment]
Sorry, didn't have a chance to review this (was scheduled for Monday), get sorted out all right???

**Comment by paf31** [comment]
Yep, @garyb and I went over it on IRC. I think this is probably the best approach for now, but we can revisit it if necessary.

**Pull request merged**